Implement predict_per_tree

hcho3 opened the pull request.

hcho3 committed:
Add predict_type parameter

hcho3 committed:
Add num_trees()

hcho3 committed:
Factor out get_device_type()

hcho3 committed:
Set up skeleton for predict_per_tree

hcho3 committed:
Rename class_count -> output_count for consistency

hcho3 committed:
Correctly set model_handle

hcho3 committed:
First implementation of predict_per_tree

hcho3 committed:
Fix bugs

hcho3 committed:
Add test coverage for vector leaf

hcho3 committed:
Fix formatting

wphicks commented on cpp/include/cuml/experimental/fil/decision_forest.hpp:
```diff
@@ -227,6 +231,7 @@ struct decision_forest {
    * larger values.
    */
   void predict(
+    predict_t predict_type,
```
Could we make this an optional parameter with a sensible default wherever possible?

chyunsu3 replied:
Done

wphicks commented on cpp/include/cuml/experimental/fil/detail/infer/gpu.cuh:
```diff
@@ -150,56 +157,89 @@ std::enable_if_t<D==raft_proto::device_type::gpu, void> infer(
-  if (output_workspace_size_bytes > max_shared_mem_per_block) {
-    throw unusable_model_exception(
-      "Model output size exceeds available shared memory"
+  index_type shared_mem_per_block{};
```
Since we're already using [AAA style](https://herbsutter.com/2013/08/12/gotw-94-solution-aaa-style-almost-always-auto/) everywhere in the new codebase and it is pretty useful for managing the heavy templating we're doing, let's stick to it here as well.

hcho3 committed:
Merge remote-tracking branch 'wphicks/fea-cpu_gpu_fil' into predict_per_tree

wphicks commented on cpp/include/cuml/experimental/fil/detail/infer/gpu.cuh:
```diff
@@ -150,56 +157,89 @@ std::enable_if_t<D==raft_proto::device_type::gpu, void> infer(
+  index_type rows_per_block_iteration{};
+  index_type output_workspace_size{};
+  auto constexpr const output_item_bytes = index_type(sizeof(typename forest_t::io_type));
+  if (predict_type == predict_t::predict) {
```
It's not immediately obvious to me why we only perform this calculation for default prediction. Don't we need to compute shared memory usage regardless?

wphicks replied:
Ah, I see that we're computing it below, but it seems like we have a lot of code duplication. Can we clean that up at all? Maybe by passing the prediction type to `compute_output_size`?

wphicks commented on cpp/include/cuml/experimental/fil/detail/infer_kernel/gpu.cuh:
```diff
@@ -104,25 +109,29 @@ infer_kernel(
 
     shared_mem.clear();
+    // Only used if predict_type == predict_t::predict
+    // Assume: output_workspace_size == 0 if predict_type is predict_per_tree or predict_leaf
```
Have we checked that using shared memory for `predict_per_tree` hurts performance for small models? It might be better to use shared memory where possible with a fallback if necessary.

wphicks commented on cpp/include/cuml/experimental/fil/detail/infer_kernel/gpu.cuh:
```diff
@@ -167,87 +176,116 @@ infer_kernel(
         }
       }
 
       __syncthreads();
```
Is this sync still necessary for `predict_per_tree`?

wphicks commented on cpp/include/cuml/experimental/fil/detail/infer_kernel/shared_memory_buffer.cuh:
```diff
@@ -91,6 +91,10 @@ struct shared_memory_buffer {
   template <typename T>
   __device__ auto* fill(index_type element_count, T value=T{}) {
+    if (element_count == 0) {
+      return static_cast<T*>(nullptr);
```
Let's try to stick to a single return statement to take advantage of RVO wherever possible.

wphicks commented on cpp/include/cuml/experimental/fil/predict_type.hpp:
```diff
@@ -0,0 +1,27 @@
+namespace experimental {
+namespace fil {
+enum class predict_t : unsigned char {
+  predict=0,
```
We might consider renaming these for a little more clarity. How would we feel about naming the whole thing `output_kind` with values `default`, `per_tree`, and `leaf_id`?

chyunsu3 replied:
I went ahead and rename the enum class to `output_kind`. Since `default` is a reserved keyword in C++, I used `default_kind` instead.

wphicks commented on python/cuml/experimental/fil/fil.pyx:
```diff
@@ -309,17 +324,75 @@ cdef class ForestInference_impl():
+                chunk_specification
+            )
+
+        self.raft_proto_handle.synchronize()
```
Why are we synchronizing here?

chyunsu3 replied:
The `predict()` function had the synchronization call, so I figured it would be necessary here too.

wphicks commented on python/cuml/experimental/fil/fil.pyx:
```diff
@@ -309,17 +324,75 @@ cdef class ForestInference_impl():
+
+        return preds
+
+    def predict_per_tree(
```
Can we consolidate the code for these two calls so that we're calling an underlying `_predict` function (e.g.) with the prediction type and `predict`/`predict_by_tree` are just thin wrappers?

chyunsu3 replied:
Done

wphicks commented on python/cuml/experimental/fil/fil.pyx:
```diff
@@ -309,17 +324,75 @@ cdef class ForestInference_impl():
+        return preds
+
+    def predict_per_tree(
+            self,
```
We should also optionally allow the user to provide an output array.

chyunsu3 replied:
Done

wphicks reviewed (commented):
Looking pretty good! Left some comments and questions around details.

hcho3 committed:
Rename enum predict_t -> output_kind

hcho3 committed:
Combine two predict functions

hcho3 committed:
Make output_type an optional parameter

hcho3 committed:
Use RVO

hcho3 committed:
Add missing preds arg

hcho3 committed:
Implement global memory fallback

cjnolet committed:
Updating cuML to use consolidated RAFT libs (#5272)

This is intended to be merged in 23.06 w/ this PR: https://github.com/rapidsai/raft/pull/1333.

Authors:
  - Corey J. Nolet (https://github.com/cjnolet)
  - Vyas Ramasubramani (https://github.com/vyasr)

Approvers:
  - Dante Gama Dessavre (https://github.com/dantegd)
  - AJ Schmidt (https://github.com/ajschmidt8)

URL: https://github.com/rapidsai/cuml/pull/5272

vyasr committed:
Stop setting package version attribute in wheels (#5285)

This PR removes modification of the `__init__.py::version` attribute that occurs during the wheel build process. See https://github.com/rapidsai/ops/issues/2592 for more information.

Authors:
  - Vyas Ramasubramani (https://github.com/vyasr)

Approvers:
  - Sevag H (https://github.com/sevagh)

URL: https://github.com/rapidsai/cuml/pull/5285

galipremsagar committed:
Fix cuml local cpp docs build (#5297)

This PR fixes the following error that shows up during local cpp docs build:
```
[1/1] Generate doxygen docs
FAILED: CMakeFiles/docs_cuml /home/whicks/proj_cuml/cuml/cpp/build/CMakeFiles/docs_cuml 
cd /home/whicks/proj_cuml/cuml/cpp/build && /home/whicks/miniconda3/envs/cuml_dev/bin/doxygen /home/whicks/proj_cuml/cuml/cpp/build/Doxyfile
error: tag HTML_HEADER: header file 'header.html' does not exist
Exiting...
```

Authors:
  - GALI PREM SAGAR (https://github.com/galipremsagar)

Approvers:
  - William Hicks (https://github.com/wphicks)
  - Corey J. Nolet (https://github.com/cjnolet)

URL: https://github.com/rapidsai/cuml/pull/5297

wphicks committed:
Provide FIL implementation for both CPU and GPU (#4890)



Authors:
  - William Hicks (https://github.com/wphicks)
  - Victor Lafargue (https://github.com/viclafargue)
  - Corey J. Nolet (https://github.com/cjnolet)
  - Carl Simon Adorf (https://github.com/csadorf)
  - Bradley Dice (https://github.com/bdice)
  - AJ Schmidt (https://github.com/ajschmidt8)

Approvers:
  - Philip Hyunsu Cho (https://github.com/hcho3)
  - Dante Gama Dessavre (https://github.com/dantegd)

URL: https://github.com/rapidsai/cuml/pull/4890

hcho3 committed:
Merge remote-tracking branch 'origin/branch-23.04' into predict_per_tree

chyunsu3 commented:
Re-submitting this to cuML repo
